In [29]:
#Importing Libraries

import numpy as np
import pandas as pd
from sklearn.linear_model import Lasso, LinearRegression, Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [30]:
# 1. Load Data
df = pd.read_csv('Student_Performance.csv')

In [31]:
# 2. Preprocessing
df['Extracurricular Activities'] = df['Extracurricular Activities'].map(
    {'Yes': 1, 'No': 0}
)
X = df.drop(columns=['Performance Index'])
y = df['Performance Index']

In [32]:
# 3. Train/Test Split (80% Train, 20% Test)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [33]:
# 4. Feature Scaling (StandardScaler is critical for Ridge and Lasso regularization)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [34]:
# 5. Model Training
# Ordinary Least Squares (Linear Regression)
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)
y_pred_lr = lr.predict(X_test_scaled)

In [35]:
# Ridge Regression (L2 Regularization)
ridge = Ridge(alpha=1.0)
ridge.fit(X_train_scaled, y_train)
y_pred_ridge = ridge.predict(X_test_scaled)

In [36]:
# Lasso Regression (L1 Regularization)
lasso = Lasso(alpha=0.01)
lasso.fit(X_train_scaled, y_train)
y_pred_lasso = lasso.predict(X_test_scaled)

In [37]:
# 6. Evaluation Function
def evaluate_model(y_true, y_pred):
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = np.sqrt(mse)
  r2 = r2_score(y_true, y_pred)
  return mae, mse, rmse, r2

In [38]:
# Compute metrics
results = {
    'Linear Regression': evaluate_model(y_test, y_pred_lr),
    'Ridge (alpha=1.0)': evaluate_model(y_test, y_pred_ridge),
    'Lasso (alpha=0.01)': evaluate_model(y_test, y_pred_lasso),
}

In [39]:
# Metric summary DataFrame
metrics_summary = pd.DataFrame(
    results, index=['MAE', 'MSE', 'RMSE', 'R²']
).T.round(4)
print(metrics_summary)

                       MAE     MSE    RMSE     R²
Linear Regression   1.6111  4.0826  2.0206  0.989
Ridge (alpha=1.0)   1.6112  4.0832  2.0207  0.989
Lasso (alpha=0.01)  1.6117  4.0872  2.0217  0.989


##Conclusion
####Linear Regression generalizes best on the test set, achieving the lowest test error (RMSE=2.0206, MAE=1.6111) and highest variance explained (R 2 =0.989). Because the dataset has a large sample size (10,000 rows) and strong true linear relationships without severe multicollinearity, the unregularized model does not overfit. Consequently, L1 (Lasso) and L2 (Ridge) penalties unnecessarily shrink the feature coefficients, leading to marginally higher test error rather than improving generalization.